### ISSUE TITLE
One type definition per file

### ISSUE THREAD

**Issue by mokagio** [iflix-letsplay/swift-style-guide#5]
The trend in the codebase seems to be "one type definition per file", which is cool.

Are we fine with this? If we are I think we should apply it rigorously (see zero-one-infinity), that is even little types and typealias should have their own file.

Thoughts?

**Comment by TimKregerIflix** [comment]
Yeah I'm in favour of this, it makes it much easier to find definitions.

**Comment by mokagio** [comment]
Yes lets make a clear "one type definition per file" rule, which should apply to tests as well.

**Comment by mokagio** [comment]
No, let's not do this and keep using the current mixed approach.

**Comment by cuteiosdev** [comment]
Generally I'm for one definition per file. There is though a problem with how we define definition. Sometimes is makes sense to an extent.

**Comment by iamariffikri** [comment]
🤔 hmm.. to be honest, my preference is more into multiple type def in a file, as long as they are related. In let say, Actions struct definition. I prefer to include all actions related to Landing into one LandingActions file. But well, if we're going with one strict rule, I dont mind, anything for team productivity! :D

**Comment by nhisyam** [comment]
In my opinion, multiple definitions in a single file is still fine as long as they are strongly related. Let's pick up something as an example:

If you take a  look at `BarButtonItem.swift` file, there are multiple `*BarButtonItem` structs being defined in a single class. Personally I think this is fine because:
- each item struct is short only 4-5 LOC
- each items are strongly related to each other.

**Comment by mokagio** [comment]
I think in this case applying the [0, 1, infinity rule](https://en.wikipedia.org/wiki/Zero_one_infinity_rule) would help taking decisions. Rather than thinking whether it makes sense or not in an occasion compared to another, we shall always put one type definition per file. Less beard scratching.

We I say with "type definition" I'm thinking of:

```swift
class|struct|enum|protocol|typealias Foo {
  ...
}
```

This would result into

### Preferred

```swift
// SomeAction.swift
struct SomeAction { }

// SomeProtocol.swift
protocol SomeProtocol {
  ...
}

// SomeType.swift
struct SomeType: SomeProtocol {
  ...
}
```

### Not Preferred

```swift
// Actions.swift
struct AnAction { }
struct ARelatedAction { }
struct AnActionThatHasSomethingToDoWithTheOtherOnes { }
```
